src/user/mod.rs

```diff
@@ -473,9 +589,209 @@ pub fn update_user(req: &mut Request) -> CargoResult<Response> {
+        user_id: user.id,
+        email: String::from(user_email),
+        verified: false,
+    };
+
+    let email_result: QueryResult<Email> =
+        insert(&new_email.on_conflict(user_id, do_update().set(&new_email)))
+            .into(emails::table)
+            .get_result(&*conn)
+            .map_err(Into::into);
+
+    let token = generate_token();
+
+    match email_result {
+        Ok(email_response) => {
+            let new_token = NewToken {
+                email_id: email_response.id,
+                token: token.clone(),
+                created_at: time::now_utc().to_timespec(),
+            };
+
+            insert(&new_token.on_conflict(
+                email_id,
+                do_update().set(&new_token),
+            )).into(tokens::table)
+                .execute(&*conn)?;
+        }
+        Err(_) => {
+            return Err(human("Error in creating token"));
+        }
+    }
+
+    let email_result = send_user_confirm_email(user_email, &user.gh_login, &token);
+    email_result.map_err(
+        |_| bad_request("Email could not be sent"),
+    )?;
+
     #[derive(Serialize)]
     struct R {
         ok: bool,
     }
     Ok(req.json(&R { ok: true }))
 }
+
+pub struct MailgunConfigVars {
+    pub smtp_login: String,
+    pub smtp_password: String,
+    pub smtp_server: String,
+}
+
+fn send_user_confirm_email(email: &str, user_name: &str, token: &str) -> CargoResult<()> {
+    // perhaps use crate lettre and heroku service Mailgun
+    // Create a URL with token string as path to send to user
+    // If user clicks on path, look email/user up in database,
+    // make sure tokens match
+
+    use dotenv::dotenv;
+    use std::env;
+    use lettre::transport::smtp::{SecurityLevel, SmtpTransportBuilder};
+    use lettre::email::EmailBuilder;
+    use lettre::transport::smtp::authentication::Mechanism;
+    use lettre::transport::smtp::SUBMISSION_PORT;
+    use lettre::transport::EmailTransport;
+    use lettre::transport::file::FileEmailTransport;
+    use std::path::Path;
+
+    dotenv().ok();
+    let mailgun_config = MailgunConfigVars {
+        smtp_login: env::var("MAILGUN_SMTP_LOGIN").unwrap_or_else(|_| String::from("Not found")),
+        smtp_password: env::var("MAILGUN_SMTP_PASSWORD").unwrap_or_else(|_| {
+            String::from("Not found")
+        }),
+        smtp_server: env::var("MAILGUN_SMTP_SERVER").unwrap_or_else(|_| String::from("Not found")),
+    };
+
+    let email = EmailBuilder::new()
+        .to(email)
+        .from(mailgun_config.smtp_login.as_str())
+        .subject("Please confirm your email address")
+        .body(
+            format!(
+                "Hello {}! Welcome to Crates.io. Please click the
+link below to verify your email address. Thank you!\n
+https://crates.io/confirm/{}",
+                user_name,
+                token
+            ).as_str(),
+        )
+        .build()
+        .expect("Failed to build confirm email message");
```

Comment:
So if this `expect` gets an `Err`, this would crash the server, which normally would make me go 😱 🙀 😱 🙀 😱 🙀  but looking at [lettre's code](https://docs.rs/lettre/0.6.2/src/lettre/email/mod.rs.html#555), the only way building an email would fail is if:

- no `from` address was set, and we're definitely calling `from` with something we control
- the `to` address is empty string, and we reject empty emails before calling `send_user_confirm_email` in `update_user` 

I'm trying to figure out if it's possible that we get `Some("")` from github and therefore would insert an empty string into the email table and then get an empty email here? I'm going to try some testing and see what happens.

This is also starting to feel a bit spooky-action-at-a-distance since the code that is ensuring the `to` email isn't empty is far away from here. This could be an issue for any email we send.... Hmmmm.... I wonder if, when the generic email functionality is extracted into a function, if we should pass the generic function strings for `to`, `from`, `subject`, and `body`, and then the generic function first checks to make sure `to` isn't an empty string and returns an `Err` if so, and then we can call `expect` and point to the line checking that should be nearby as proof that the `expect` won't fail?